cpp/src/binaryop/compiled/equality_ops.cu

```diff
@@ -14,44 +14,75 @@
  * limitations under the License.
  */
 
-#include "binary_ops.cuh"
+#include <binaryop/compiled/binary_ops.cuh>
+#include <binaryop/compiled/struct_binary_ops.cuh>
+
+#include <cudf/detail/structs/utilities.hpp>
+#include <cudf/table/row_operators.cuh>
 
 namespace cudf::binops::compiled {
-void dispatch_equality_op(mutable_column_device_view& outd,
-                          column_device_view const& lhsd,
-                          column_device_view const& rhsd,
+void dispatch_equality_op(mutable_column_view& out,
+                          column_view const& lhs,
+                          column_view const& rhs,
                           bool is_lhs_scalar,
                           bool is_rhs_scalar,
                           binary_operator op,
                           rmm::cuda_stream_view stream)
 {
-  auto common_dtype = get_common_type(outd.type(), lhsd.type(), rhsd.type());
-
-  // Execute it on every element
+  if (is_struct(lhs.type()) && is_struct(rhs.type())) {
+    auto const nullability =
+      structs::detail::contains_null_structs(lhs) || structs::detail::contains_null_structs(rhs)
+        ? structs::detail::column_nullability::FORCE
+        : structs::detail::column_nullability::MATCH_INCOMING;
+    auto const lhs_flattened =
+      structs::detail::flatten_nested_columns(table_view{{lhs}}, {}, {}, nullability);
+    auto const rhs_flattened =
+      structs::detail::flatten_nested_columns(table_view{{rhs}}, {}, {}, nullability);
+    auto d_lhs = table_device_view::create(lhs_flattened);
+    auto d_rhs = table_device_view::create(rhs_flattened);
 
-  if (common_dtype) {
-    if (op == binary_operator::EQUAL) {
-      for_each(stream,
-               outd.size(),
-               binary_op_device_dispatcher<ops::Equal>{
-                 *common_dtype, outd, lhsd, rhsd, is_lhs_scalar, is_rhs_scalar});
-    } else if (op == binary_operator::NOT_EQUAL) {
-      for_each(stream,
-               outd.size(),
-               binary_op_device_dispatcher<ops::NotEqual>{
-                 *common_dtype, outd, lhsd, rhsd, is_lhs_scalar, is_rhs_scalar});
-    }
+    if (op == binary_operator::EQUAL || op == binary_operator::NOT_EQUAL)
+      detail::struct_compare(
+        out,
+        row_equality_comparator{
+          nullate::DYNAMIC{has_nested_nulls(lhs_flattened) || has_nested_nulls(rhs_flattened)},
+          *d_lhs,
+          *d_rhs},
+        is_lhs_scalar,
+        is_rhs_scalar,
+        op == binary_operator::NOT_EQUAL,
```

Comment:
Does this work correctly for floating point `NaN==NaN` is false and `NaN!=NaN` is false ?

Reply:
`NaN!=NaN` should return true.
Since we are flipping output for >=, <, and <=, >, Better to confirm if it behaves as described.

https://github.com/rapidsai/cudf/issues/4760 is the discussion about NaN behaviors


@jrhemstad  we should follow C++ NaN behavior right?

Reply:
Does not match the C++ standard, it's giving results similar to the the expected spark behavior but `NaN>NaN` is returning false. Looking into how to fix this issue.

Reply:
We had similar discussion about different results of `NaN == NaN` in `lists::drop_list_duplicates`. The only way out is to have an option allowing to choose the desired behavior (https://github.com/rapidsai/cudf/blob/branch-22.02/cpp/include/cudf/types.hpp#L151-L154).

Reply:
With this in mind, is the logical next step to add `nan_equality` and a `nan_comparison`  configurations to the `element_equality_comparator`, `element_relational_comparator`, `row_equality_comparator`, and `row_lexicographic_comparator`.